base/pkg/read.jl

```diff
@@ -2,15 +2,70 @@
 
 module Read
 
-import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir
+import ...LibGit2, ..Cache, ..Reqs, ...Pkg.PkgError, ..Dir, ..isdevmetadata
 using ..Types
 
+devmetapkgpath(s::AbstractString) = joinpath(uppercase(string(s[1])), s * ".versions")
+
 readstrip(path...) = strip(readstring(joinpath(path...)))
 
-url(pkg::AbstractString) = readstrip(Dir.path("METADATA"), pkg, "url")
-sha1(pkg::AbstractString, ver::VersionNumber) = readstrip(Dir.path("METADATA"), pkg, "versions", string(ver), "sha1")
+function url(pkg::AbstractString, dev::Bool = false)
+    if dev
+        chomp(readline(joinpath(Dir.path("METADATA"), devmetapkgpath(pkg))))
+    else
+        readstrip(Dir.path("METADATA"), pkg, "url")
+    end
+end
+
+# Should only be called when cache is guaranteed to be updated
```

Comment:
sounds like. Pkg is probably already racy until we implement file locking (can't remeber the issue number for that) so wouldn't worry too much about the race condition here.

if we use bare repos without materializing a visible clone to the filesystem, then it gets quite difficult to debug anything that goes wrong or have a manual fallback. anything bare should have an editable mode, when that is active the filesystem should be the primary source of data that gets used (with the git version only as a cache which needs to have no local  modifications to be valid). when not in editable mode, it should be obviously visible - no filesystem full copy at all, otherwise it's going to be extremely confusing and error prone to keep track of.